api/utils/sshutils/ssh.go

```diff
@@ -250,3 +260,263 @@ func KeysEqual(ak, bk ssh.PublicKey) bool {
+// The file format is described here: https://the.earth.li/~sgtatham/putty/0.76/htmldoc/AppendixC.html#ppk
+func ConvertToPPK(priv []byte, pub []byte) ([]byte, error) {
+	// decode the private key from PEM format and extract the exponents
+	privateKeyPemBlock, rest := pem.Decode(priv)
+	if len(rest) > 0 {
+		return nil, trace.Wrap(fmt.Errorf("failed to decode private key, %v bytes left over", len(rest)))
+	}
+	privateKey, err := x509.ParsePKCS1PrivateKey(privateKeyPemBlock.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to parse private key: %T", err))
+	}
+
+	// https://the.earth.li/~sgtatham/putty/0.76/htmldoc/AppendixC.html#ppk
+	// RSA keys are stored using an algorithm-name of ‘ssh-rsa’. (Keys stored like this are also used by the updated RSA signature schemes that use
+	// hashes other than SHA-1. The public key data has already provided the key modulus and the public encoding exponent. The private data stores:
+	// mpint: the private decoding exponent of the key.
+	// mpint: one prime factor p of the key.
+	// mpint: the other prime factor q of the key. (RSA keys stored in this format are expected to have exactly two prime factors.)
+	// mpint: the multiplicative inverse of q modulo p.
+	ppkPrivateKey := new(bytes.Buffer)
+
+	// mpint: the private decoding exponent of the key.
+	// this is known as 'D'
+	mpintD, err := getRFC4251Mpint(privateKey.D)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(ppkPrivateKey, binary.BigEndian, mpintD)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to write D to buffer: %T", err))
+	}
+
+	// mpint: one prime factor p of the key.
+	// this is known as 'P'
+	// the RSA standard dictates that P > Q
+	// for some reason what PuTTY names 'P' is Primes[1] to Go, and what PuTTY names 'Q' is Primes[0] to Go
+	P := privateKey.Primes[1]
+	mpintP, err := getRFC4251Mpint(P)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(ppkPrivateKey, binary.BigEndian, mpintP)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to write P to buffer: %T", err))
+	}
+
+	// mpint: the other prime factor q of the key. (RSA keys stored in this format are expected to have exactly two prime factors.)
+	// this is known as 'Q'
+	Q := privateKey.Primes[0]
+	mpintQ, err := getRFC4251Mpint(Q)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(ppkPrivateKey, binary.BigEndian, mpintQ)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to write Q to buffer: %T", err))
+	}
+
+	// mpint: the multiplicative inverse of q modulo p.
+	// this is known as 'iqmp'
+	iqmp := new(big.Int).ModInverse(Q, P)
+	mpintIQMP, err := getRFC4251Mpint(iqmp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(ppkPrivateKey, binary.BigEndian, mpintIQMP)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to write iqmp to buffer: %T", err))
```

Comment:
`%T` is going to print the type of `err`, which is probably not useful in an error message (though as mentioned above, this is guaranteed not to error)

Reply:
Got it. I've removed a lot of these anyway, and changed the ones that remain to print the error as `%v` instead.